core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -78,34 +78,19 @@ sealed abstract class Parser0[+A] { self: Product =>
       )
   }
 
+  private[this] lazy val allParser = (this ~ Parser.end)
+
   /** Attempt to parse all of the input `str` into an `A` value.
     *
     * This method will return a failure unless all of `str` is consumed during parsing.
     *
     * `p.parseAll(s)` is equivalent to `(p <* Parser.end).parse(s).map(_._2)`.
     */
-  final def parseAll(str: String): Either[Parser.Error, A] = {
```

Comment:
this old implementation was actually hiding some error messages which the optimization of `a.? ~ b` discovered. Replaced with a direct implementation.